src/murfey/workflows/clem/register_preprocessing_results.py

```diff
@@ -0,0 +1,397 @@
+"""
+Functions to process the requests received by Murfey related to the CLEM workflow.
+
+The CLEM-related file registration API endpoints can eventually be moved here, since
+the file registration processes all take place on the server side only.
+"""
+
+from __future__ import annotations
+
+import json
+import logging
+import re
+import traceback
+from pathlib import Path
+from typing import Optional, Type, Union
+
+from sqlalchemy.exc import NoResultFound
+from sqlmodel import Session, select
+
+from murfey.util.config import get_machine_config
+from murfey.util.db import (
+    CLEMImageMetadata,
+    CLEMImageSeries,
+    CLEMImageStack,
+    CLEMLIFFile,
+    CLEMTIFFFile,
+)
+from murfey.util.db import Session as MurfeySession
+from murfey.util.models import LIFPreprocessingResult, TIFFPreprocessingResult
+
+logger = logging.getLogger("murfey.workflows.clem.register_results")
+
+
+def _validate_and_sanitise(
+    file: Path,
+    session_id: int,
+    db: Session,
+) -> Path:
+    """
+    Performs validation and sanitisation on the incoming file paths, ensuring that
+    no forbidden characters are present and that the the path points only to allowed
+    sections of the file server.
+
+    Returns the file path as a sanitised string that can be converted into a Path
+    object again.
+
+    NOTE: Due to the instrument name query, 'db' now needs to be passed as an
+    explicit variable to this function from within a FastAPI endpoint, as using the
+    instance that was imported directly won't load it in the correct state.
+    """
+
+    valid_file_types = (
+        ".lif",
+        ".tif",
+        ".tiff",
+        ".xlif",
+        ".xml",
+    )
+
+    # Resolve symlinks and directory changes to get full file path
+    full_path = Path(file).resolve()
+
+    # Use machine configuration to validate which file base paths are accepted from
+    instrument_name = (
+        db.exec(select(MurfeySession).where(MurfeySession.id == session_id))
+        .one()
+        .instrument_name
+    )
+    machine_config = get_machine_config(instrument_name=instrument_name)[
+        instrument_name
+    ]
+    rsync_basepath = machine_config.rsync_basepath
+    try:
+        base_path = list(rsync_basepath.parents)[-2].as_posix()
```

Comment:
Not sure I understand why you're finding the parents of the rsync_basepath? Shouldn't the file always be relative to the rsync_basepath?